### PR TITLE
ReactDOMComponentTree

### PR DESCRIPTION
**Pull request by spicyj**
New module! With this, we have a new module that uses the component tree to store DOM nodes. Most of the logic in this file relates to markup adoption -- if we were to drop that (along with server rendering) this file could just be a `return inst._nativeNode;`.

This works with useCreateElement only because we need to reference each node as it's created.

Events is now the only thing using hierarchical IDs and `ReactMount.getNode` -- I'll use `ReactDOMComponentTree.getInstanceFromNode` to make that work.

### CODE REVIEW

File: src/renderers/shared/reconciler/__tests__/ReactEmptyComponent-test.js [hunk 1]
```diff
@@ -81,9 +85,7 @@ describe('ReactEmptyComponent', function() {
     );
   });
 
-  it('should be able to switch between rendering null and a normal tag', () => {
-    spyOn(console, 'log');
-
+  it.only('should be able to switch between rendering null and a normal tag', () => {
```
Comment (gaearon) [inline src/renderers/shared/reconciler/__tests__/ReactEmptyComponent-test.js]: it.only

Comment (spicyj) [inline src/renderers/shared/reconciler/__tests__/ReactEmptyComponent-test.js]: Thanks, fixed.

**Comment by gaearon** [comment]
What is the big picture here? Getting rid of `data-reactid` eventually? Better resilience against evil browser plugins?

**Comment by spicyj** [comment]
Getting rid of hierarchical IDs (replacing them with numbers, probably) and the giant hash map we have in ReactMount. We can probably kill data-reactid entirely when not using server-side rendering.

**Comment by jimfb** [comment]
My hope is that we can kill `data-reactid` entirely **especially** when using server-side rendering.  Rehydrating a markup tree does not benefit from having `data-reactid` afik.

File: src/renderers/dom/shared/__tests__/ReactDOMComponent-test.js [hunk 2]
```diff
@@ -52,7 +52,7 @@ describe('ReactDOMComponent', function() {
       var stubStyle = container.firstChild.style;
 
       // set initial style
-      var setup = {display: 'block', left: '1', top: 2, fontFamily: 'Arial'};
+      var setup = {display: 'block', left: '1px', top: 2, fontFamily: 'Arial'};
```
Comment (sebmarkbage) [inline src/renderers/dom/shared/__tests__/ReactDOMComponent-test.js]: What's this about? Behavior change?

Comment (jimfb) [inline src/renderers/dom/shared/__tests__/ReactDOMComponent-test.js]: Styles may be unitless if they are integers, but unitless strings are deprecated.  This looks fine to me.

Comment (spicyj) [inline src/renderers/dom/shared/__tests__/ReactDOMComponent-test.js]: @jimfb recently merged a PR that adds a warning for this case but it only checked the create case, not the update case which we use for createElement. I fixed that so the test for the warning passes with useCreateElement and fixed this test to not warn.

**Comment by spicyj** [comment]
@jimfb I'm operating under the assumption that browser extensions will insert extra nodes into our markup and we need to skip over them.

**Comment by sebmarkbage** [comment]
Make sure you do your react-native homework

**Comment by syranide** [comment]
> My hope is that we can kill data-reactid entirely especially when using server-side rendering. Rehydrating a markup tree does not benefit from having data-reactid afik.

@jimfb FYI, we can already kill data-reactid for server-rendered markup, it's rather trivial as long as you're ok with doing a fast pass over all the nodes once when starting the client.

**Comment by spicyj** [comment]
We can do that trivially here too by changing precacheChildNodes to not look at IDs and just match up the nodes – but I'm worried about finding extra nodes.

**Comment by syranide** [comment]
@spicyj We can always mark the nodes with just `data-react`, but leave out the ID. That should give you most of the benefits, while also tolerating foreign nodes. Personally I think we should just be super strict and simply not allow unexpected nodes, they are crashes waiting to happen.

**Comment by spicyj** [comment]
That's true, but I don't think including a numeric ID (`data-reactid="17"`) is much overhead over that.

**Comment by jimfb** [comment]
> Personally I think we should just be super strict and simply not allow unexpected nodes, they are crashes waiting to happen.

I tend to agree - if we find a foreign node while restoring a SSR DOM tree, we can just kill it with fire.

**Comment by syranide** [comment]
> That's true, but I don't think including a numeric ID (data-reactid="17") is much overhead over that.

@spicyj `data-react` compresses trivially to virtually nothing, but yeah the overhead of the ID should be minimal too. Both are miles better than the current reactids anyway.

**Comment by spicyj** [comment]
@jimfb @sebmarkbage The last two commits add events support so we don't rely on hierarchical IDs for events either. (Sorry for piling it all into the same PR – but github sucks at stacking PRs and I didn't want to merge this to master quite yet.) Either of you want to take a look? I'm a little nervous but all the FB test flows worked the first time with this applied.

**Comment by spicyj** [comment]
Did a quick benchmark of rendering [a PE-like hierarchy](https://gist.github.com/spicyj/36f063272e4849c0e0a3) with Electron (Chrome) with a cold and warm JIT using my perf-counters as the timestamp.

With a warm JIT:

This PR is 3.3% faster than master.
Switching to numerical IDs (will be a separate PR, not quite done yet) brings us to 7.9% faster than master.
Numerical IDs with createElement instead of HTML generation is on par (also 7.9%).
Numerical IDs with createElement and not setting `data-reactid` at all (since we don't need it) gives another 1.4% boost for a total of 9.1%.

With a cold JIT, createElement looks a little bit slower (but this PR helps more compared to master!).

Our deltas of
[3.3%, 7.9%, 7.9%, 9.1%] with a warm JIT are
[4.9%, 9.6%, 8.3%, 8.9%] with a cold JIT.

Not sure if there are things we can do to improve the speed of createElement.

**Comment by syranide** [comment]
@spicyj Just for context, I assume those are React rendering times + `innerHTML` but do not include browser rendering/painting?

**Comment by spicyj** [comment]
```
var a = now();
React.render(React.createElement(Benchmark), container);
container.offsetHeight;
var b = now();
```

There's probably some painting involved. I intentionally forced layout: I wanted to make sure all the markup was parsed to make sure the HTML mode wasn't doing stuff lazily. Any ideas on how to keep that guarantee while avoiding painting (and layout)?

**Comment by syranide** [comment]
@spicyj I'd assume that would just force layouting, realistically rendering/painting should only happen after JS execution ends. I'm pretty sure `innerHTML` is not deferred as you can measure differences in time by the number of tags/attributes/etc present. It probably doesn't really make sense to defer it in practice either I would think.

**Comment by spicyj** [comment]
You're probably right. I tried it again without the forced layout and it's about 18.6ms faster on all five numbers, so the percentages are proportionally the same but go up a little because of a smaller base. Cold JIT, no layout: [5.5%, 10.8%, 9.2%, 10.0%].

**Comment by spicyj** [comment]
Numerical IDs are in #5205. Also, devtools don't work with this because they currently rely on `ReactMount.getNode`.

**Comment by spicyj** [comment]
Merged in #5205.

**Pull request closed without merging**